OpenRA.Mods.Common/Traits/World/VariedCostManager.cs

```diff
@@ -0,0 +1,60 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2021 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System.Collections.Generic;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Attach this to the world actor.")]
+	public class VariedCostManagerInfo : TraitInfo
+	{
+		[Desc("Interval between new pricings measured in ticks.")]
+		public readonly int AdjustmentDelay = 1000;
+
+		[Desc("Range of percentage modifiers to apply.")]
+		public readonly int2 Multiplier = new int2(75, 100);
+
+		public override object Create(ActorInitializer init) { return new VariedCostManager(this); }
+	}
+
+	public class VariedCostManager : ITick
+	{
+		public readonly Dictionary<string, int> CachedCostPercentage = new Dictionary<string, int>();
+
+		readonly VariedCostManagerInfo info;
+
+		int tick;
+
+		public VariedCostManager(VariedCostManagerInfo info)
+		{
+			this.info = info;
+		}
+
+		void ITick.Tick(Actor self)
+		{
+			if (--tick <= 0)
+			{
+				foreach (var actor in self.World.Map.Rules.Actors)
+				{
+					var variedCostMultiplierInfo = actor.Value.TraitInfoOrDefault<VariedCostMultiplierInfo>();
```

Comment:
These should be cacheable, so you'd only need a dictionary with key (actor name or group) and value (default price), which this `foreach` can go over. The `CachedCostPecentage` dictionary stays, of course.

Reply:
I cached them like we do in the bot modules now.